torch/fx/node.py

```diff
@@ -393,6 +393,16 @@ def maybe_replace_node(n : Node) -> Node:
         assert len(self.users) == 0
         return to_process
 
+    def is_impure(self):
+        """
+        Returns whether this op is impure, i.e. if its op is a placeholder or output.
+
+        Returns:
+
+            bool: If the op is impure or not.
+        """
+        return self.op in {"placeholder", "output"}
```

Comment:
I recommend doing something like (after https://github.com/pytorch/pytorch/pull/52358 is landed):

```
# At global scope
_side_effectful_modules : Set[Type] = {}
_side_effectful_functions : Set[Callable] = {}

<...>

    def is_impure(self):
        if self.op in {"placeholder", "output"}:
          return True

        if self.op == 'call_function':
          return self.target in _side_effectful_functions

        if self.graph.owning_module and self.op == 'call_module':
            # This could be cleaned up if that PR changes `has_submodule` to
            # `get_submodule`
            assert self.graph.owning_module.has_submodule(node.target)
            mod = self.graph.owning_module
            for atom in self.target.split('.'):
                mod = getattr(md, atom)

            # Check for known side-effectful Modules.
            # Note we could also delete this if we just mark
            # those modules as `_is_impure`
            if type(mod) in _side_effectful_modules:
                return True

            # Check if the Module has an `_is_impure` attribute
            # If it's true, let's take their word for it
            return getattr(mod, '_is_impure', False)

        return False

```

Reply:
For checking if functions have side effects, it might also be possible to pull out the TorchScript schema for a given PyTorch op and check if there is aliasing. e.g. I pull out the schema for an op here: https://github.com/pytorch/pytorch/pull/49775/files#diff-03652431ccd14a2e043861738619b0b73977b3ce531f28b10946b1aacdc7456fR90 and querying the aliasing info shouldn't be too hard.

Having a hardcoded registry at first might be OK though, we can explore adding this afterward